LXD fails to start after update to snap stable 3.6 (9206)

Previously we were tracking stable, I think it was 3.5, though some of them may have been on candidate up until a few days ago.

Ubuntu Bionic, kernel is 4.15.0-1021-gcp, lxd is backed on ZFS.

## syslog output

```
Oct 14 21:20:14 server-name lxd.daemon[21959]: => Starting LXD
Oct 14 21:20:45 server-name lxd.daemon[21959]: lvl=warn msg="Failed to rollback transaction after error (failed to apply update 11: Failed to add project_id column: driver: bad connection): driver: bad connection" t=2018-10-14T21:20:45+0000
Oct 14 21:20:56 server-name lxd.daemon[21959]: lvl=eror msg="Failed to start the daemon: failed to open cluster database: failed to ensure schema: failed to begin transaction: failed to create dqlite connection: no available dqlite leader server found" t=2018-10-14T21:20:56+0000
Oct 14 21:20:56 server-name lxd.daemon[21959]: lvl=eror msg="Dqlite: aborting (fd=23 state=body msg=response write error: broken pipe (EPIPE))" t=2018-10-14T21:20:56+0000
Oct 14 21:20:56 server-name lxd.daemon[21959]: Error: failed to open cluster database: failed to ensure schema: failed to begin transaction: failed to create dqlite connection: no available dqlite leader server found
Oct 14 21:20:56 server-name lxd.daemon[21959]: => LXD failed to start
```

Per instructions in another issue, stopped lxd units, ran `lxd --debug --group lxd`:

```
DBUG[10-14|21:37:43] Connecting to a local LXD over a Unix socket 
DBUG[10-14|21:37:43] Sending request to LXD                   etag= method=GET url=http://unix.socket/1.0
INFO[10-14|21:37:44] LXD 3.6 is starting in normal mode       path=/var/snap/lxd/common/lxd
INFO[10-14|21:37:44] Kernel uid/gid map: 
INFO[10-14|21:37:44]  - u 0 0 4294967295 
INFO[10-14|21:37:44]  - g 0 0 4294967295 
INFO[10-14|21:37:44] Configured LXD uid/gid map: 
INFO[10-14|21:37:44]  - u 0 1000000 1000000000 
INFO[10-14|21:37:44]  - g 0 1000000 1000000000 
DBUG[10-14|21:37:44] Failed to set NETLINK_DUMP_STRICT_CHK socket option 
DBUG[10-14|21:37:44] Running kernel does not support netnsid-based network retrieval 
DBUG[10-14|21:37:44] Running kernel does not support uevent injection 
DBUG[10-14|21:37:44] System supports unprivileged file capabilities 
INFO[10-14|21:37:44] Initializing local database 
INFO[10-14|21:37:44] Initializing database gateway 
INFO[10-14|21:37:44] Start database node                      address= id=1
INFO[10-14|21:37:44] Raft: Restored from snapshot 1-25819-1539515095227 
INFO[10-14|21:37:48] Raft: Initial configuration (index=1): [{Suffrage:Voter ID:1 Address:0}] 
INFO[10-14|21:37:48] Raft: Node at 0 [Leader] entering Leader state 
INFO[10-14|21:37:48] Dqlite: starting event loop 
DBUG[10-14|21:37:48] Dqlite: accepting connections 
INFO[10-14|21:37:48] LXD isn't socket activated 
DBUG[10-14|21:37:48] Connecting to a local LXD over a Unix socket 
DBUG[10-14|21:37:48] Sending request to LXD                   etag= method=GET url=http://unix.socket/1.0
DBUG[10-14|21:37:48] Detected stale unix socket, deleting 
INFO[10-14|21:37:48] Starting /dev/lxd handler: 
INFO[10-14|21:37:48]  - binding devlxd socket                 socket=/var/snap/lxd/common/lxd/devlxd/sock
INFO[10-14|21:37:48] REST API daemon: 
INFO[10-14|21:37:48]  - binding Unix socket                   socket=/var/snap/lxd/common/lxd/unix.socket
INFO[10-14|21:37:48]  - binding TCP socket                    socket=10.240.0.92:8443
INFO[10-14|21:37:48] Initializing global database 
INFO[10-14|21:37:48] Dqlite: handling new connection (fd=20) 
INFO[10-14|21:37:48] Dqlite: connected address=0 attempt=0 
ERROR: can't write page 2182 of file db.bin which has only 2180 pages
panic: checkpoint: disk I/O error

trace:
2018-10-14 21:37:44.85310: fsm: restore=25819 start
2018-10-14 21:37:44.85310: fsm: restore=25819 database size: 8929280
2018-10-14 21:37:44.85951: fsm: restore=25819 wal size: 3073552
2018-10-14 21:37:44.86159: fsm: restore=25819 filename: db.bin
2018-10-14 21:37:44.86160: fsm: restore=25819 transaction ID: 
2018-10-14 21:37:44.87305: fsm: restore=25819 open follower: db.bin
2018-10-14 21:37:44.87436: fsm: restore=25819 done
2018-10-14 21:37:48.97013: fsm: term=1 index=25820 cmd=frames txn=25819 pages=3 commit=1 start
2018-10-14 21:37:48.97028: fsm: term=1 index=25820 cmd=frames txn=25819 pages=3 commit=1 unregister txn
2018-10-14 21:37:48.97028: fsm: term=1 index=25820 cmd=frames txn=25819 pages=3 commit=1 done
2018-10-14 21:37:48.97029: fsm: term=1 index=25821 cmd=frames txn=25820 pages=10 commit=1 start
2018-10-14 21:37:48.97036: fsm: term=1 index=25821 cmd=frames txn=25820 pages=10 commit=1 unregister txn
2018-10-14 21:37:48.97036: fsm: term=1 index=25821 cmd=frames txn=25820 pages=10 commit=1 done
2018-10-14 21:37:48.97053: fsm: term=1 index=25822 cmd=frames txn=25821 pages=2 commit=1 start
2018-10-14 21:37:48.97057: fsm: term=1 index=25822 cmd=frames txn=25821 pages=2 commit=1 unregister txn
2018-10-14 21:37:48.97059: fsm: term=1 index=25822 cmd=frames txn=25821 pages=2 commit=1 done
2018-10-14 21:37:48.97060: fsm: term=1 index=25823 cmd=frames txn=25822 pages=2 commit=1 start
2018-10-14 21:37:48.97063: fsm: term=1 index=25823 cmd=frames txn=25822 pages=2 commit=1 unregister txn
2018-10-14 21:37:48.97063: fsm: term=1 index=25823 cmd=frames txn=25822 pages=2 commit=1 done
2018-10-14 21:37:48.97070: fsm: term=1 index=25824 cmd=frames txn=25823 pages=3 commit=1 start
2018-10-14 21:37:48.97084: fsm: term=1 index=25824 cmd=frames txn=25823 pages=3 commit=1 unregister txn
2018-10-14 21:37:48.97084: fsm: term=1 index=25824 cmd=frames txn=25823 pages=3 commit=1 done
2018-10-14 21:37:48.97133: fsm: term=1 index=25825 cmd=frames txn=25824 pages=18 commit=1 start
2018-10-14 21:37:48.97163: fsm: term=1 index=25825 cmd=frames txn=25824 pages=18 commit=1 unregister txn
2018-10-14 21:37:48.97164: fsm: term=1 index=25825 cmd=frames txn=25824 pages=18 commit=1 done
2018-10-14 21:37:48.97165: fsm: term=1 index=25826 cmd=frames txn=25825 pages=3 commit=1 start
2018-10-14 21:37:48.97171: fsm: term=1 index=25826 cmd=frames txn=25825 pages=3 commit=1 unregister txn
2018-10-14 21:37:48.97171: fsm: term=1 index=25826 cmd=frames txn=25825 pages=3 commit=1 done
2018-10-14 21:37:48.97175: fsm: term=1 index=25827 cmd=frames txn=25826 pages=17 commit=1 start
2018-10-14 21:37:48.97200: fsm: term=1 index=25827 cmd=frames txn=25826 pages=17 commit=1 unregister txn
2018-10-14 21:37:48.97200: fsm: term=1 index=25827 cmd=frames txn=25826 pages=17 commit=1 done
2018-10-14 21:37:48.97212: fsm: term=1 index=25828 cmd=frames txn=25827 pages=9 commit=1 start
2018-10-14 21:37:48.97224: fsm: term=1 index=25828 cmd=frames txn=25827 pages=9 commit=1 unregister txn
2018-10-14 21:37:48.97228: fsm: term=1 index=25828 cmd=frames txn=25827 pages=9 commit=1 done
2018-10-14 21:37:48.97230: fsm: term=1 index=25829 cmd=frames txn=25828 pages=2 commit=1 start
2018-10-14 21:37:48.97253: fsm: term=1 index=25829 cmd=frames txn=25828 pages=2 commit=1 unregister txn
2018-10-14 21:37:48.97254: fsm: term=1 index=25829 cmd=frames txn=25828 pages=2 commit=1 done
2018-10-14 21:37:48.97288: fsm: term=1 index=25830 cmd=frames txn=25829 pages=3 commit=1 start
2018-10-14 21:37:48.97295: fsm: term=1 index=25830 cmd=frames txn=25829 pages=3 commit=1 unregister txn
2018-10-14 21:37:48.97295: fsm: term=1 index=25830 cmd=frames txn=25829 pages=3 commit=1 done
2018-10-14 21:37:48.97298: fsm: term=1 index=25831 cmd=frames txn=25830 pages=10 commit=1 start
2018-10-14 21:37:48.97305: fsm: term=1 index=25831 cmd=frames txn=25830 pages=10 commit=1 unregister txn
2018-10-14 21:37:48.97305: fsm: term=1 index=25831 cmd=frames txn=25830 pages=10 commit=1 done
2018-10-14 21:37:48.97306: fsm: term=1 index=25832 cmd=frames txn=25831 pages=2 commit=1 start
2018-10-14 21:37:48.97310: fsm: term=1 index=25832 cmd=frames txn=25831 pages=2 commit=1 unregister txn
2018-10-14 21:37:48.97310: fsm: term=1 index=25832 cmd=frames txn=25831 pages=2 commit=1 done
2018-10-14 21:37:48.97314: fsm: term=1 index=25833 cmd=frames txn=25832 pages=2 commit=1 start
2018-10-14 21:37:48.97317: fsm: term=1 index=25833 cmd=frames txn=25832 pages=2 commit=1 unregister txn
2018-10-14 21:37:48.97317: fsm: term=1 index=25833 cmd=frames txn=25832 pages=2 commit=1 done
2018-10-14 21:37:48.97318: fsm: term=1 index=25834 cmd=frames txn=25833 pages=3 commit=1 start
2018-10-14 21:37:48.97323: fsm: term=1 index=25834 cmd=frames txn=25833 pages=3 commit=1 unregister txn
2018-10-14 21:37:48.97323: fsm: term=1 index=25834 cmd=frames txn=25833 pages=3 commit=1 done
2018-10-14 21:37:48.97327: fsm: term=1 index=25835 cmd=frames txn=25834 pages=15 commit=1 start
2018-10-14 21:37:48.97338: fsm: term=1 index=25835 cmd=frames txn=25834 pages=15 commit=1 unregister txn
2018-10-14 21:37:48.97338: fsm: term=1 index=25835 cmd=frames txn=25834 pages=15 commit=1 done
2018-10-14 21:37:48.97341: fsm: term=1 index=25836 cmd=frames txn=25835 pages=6 commit=1 start
2018-10-14 21:37:48.97348: fsm: term=1 index=25836 cmd=frames txn=25835 pages=6 commit=1 unregister txn
2018-10-14 21:37:48.97348: fsm: term=1 index=25836 cmd=frames txn=25835 pages=6 commit=1 done
2018-10-14 21:37:48.97350: fsm: term=1 index=25837 cmd=frames txn=25836 pages=15 commit=1 start
2018-10-14 21:37:48.97361: fsm: term=1 index=25837 cmd=frames txn=25836 pages=15 commit=1 unregister txn
2018-10-14 21:37:48.97361: fsm: term=1 index=25837 cmd=frames txn=25836 pages=15 commit=1 done
2018-10-14 21:37:48.97367: fsm: term=1 index=25838 cmd=frames txn=25837 pages=9 commit=1 start
2018-10-14 21:37:48.97377: fsm: term=1 index=25838 cmd=frames txn=25837 pages=9 commit=1 unregister txn
2018-10-14 21:37:48.97377: fsm: term=1 index=25838 cmd=frames txn=25837 pages=9 commit=1 done
2018-10-14 21:37:48.97379: fsm: term=1 index=25839 cmd=frames txn=25838 pages=3 commit=1 start
2018-10-14 21:37:48.97425: fsm: term=1 index=25839 cmd=frames txn=25838 pages=3 commit=1 unregister txn
2018-10-14 21:37:48.97425: fsm: term=1 index=25839 cmd=frames txn=25838 pages=3 commit=1 done
2018-10-14 21:37:48.97428: fsm: term=1 index=25840 cmd=frames txn=25839 pages=18 commit=1 start
2018-10-14 21:37:48.97441: fsm: term=1 index=25840 cmd=frames txn=25839 pages=18 commit=1 unregister txn
2018-10-14 21:37:48.97441: fsm: term=1 index=25840 cmd=frames txn=25839 pages=18 commit=1 done
2018-10-14 21:37:48.97453: fsm: term=1 index=25841 cmd=frames txn=25840 pages=3 commit=1 start
2018-10-14 21:37:48.97460: fsm: term=1 index=25841 cmd=frames txn=25840 pages=3 commit=1 unregister txn
2018-10-14 21:37:48.97460: fsm: term=1 index=25841 cmd=frames txn=25840 pages=3 commit=1 done
2018-10-14 21:37:48.97482: fsm: term=1 index=25842 cmd=frames txn=25841 pages=13 commit=1 start
2018-10-14 21:37:48.97492: fsm: term=1 index=25842 cmd=frames txn=25841 pages=13 commit=1 unregister txn
2018-10-14 21:37:48.97492: fsm: term=1 index=25842 cmd=frames txn=25841 pages=13 commit=1 done
2018-10-14 21:37:48.97494: fsm: term=1 index=25843 cmd=frames txn=25842 pages=9 commit=1 start
2018-10-14 21:37:48.97501: fsm: term=1 index=25843 cmd=frames txn=25842 pages=9 commit=1 unregister txn
2018-10-14 21:37:48.97501: fsm: term=1 index=25843 cmd=frames txn=25842 pages=9 commit=1 done
2018-10-14 21:37:48.97503: fsm: term=1 index=25844 cmd=frames txn=25843 pages=2 commit=1 start
2018-10-14 21:37:48.97505: fsm: term=1 index=25844 cmd=frames txn=25843 pages=2 commit=1 unregister txn
2018-10-14 21:37:48.97505: fsm: term=1 index=25844 cmd=frames txn=25843 pages=2 commit=1 done
2018-10-14 21:37:48.97510: fsm: term=1 index=25845 cmd=frames txn=25844 pages=2 commit=1 start
2018-10-14 21:37:48.97514: fsm: term=1 index=25845 cmd=frames txn=25844 pages=2 commit=1 unregister txn
2018-10-14 21:37:48.97514: fsm: term=1 index=25845 cmd=frames txn=25844 pages=2 commit=1 done
2018-10-14 21:37:48.97515: fsm: term=1 index=25846 cmd=frames txn=25845 pages=3 commit=1 start
2018-10-14 21:37:48.97518: fsm: term=1 index=25846 cmd=frames txn=25845 pages=3 commit=1 unregister txn
2018-10-14 21:37:48.97518: fsm: term=1 index=25846 cmd=frames txn=25845 pages=3 commit=1 done
2018-10-14 21:37:48.97520: fsm: term=1 index=25847 cmd=frames txn=25846 pages=10 commit=1 start
2018-10-14 21:37:48.97529: fsm: term=1 index=25847 cmd=frames txn=25846 pages=10 commit=1 unregister txn
2018-10-14 21:37:48.97529: fsm: term=1 index=25847 cmd=frames txn=25846 pages=10 commit=1 done
2018-10-14 21:37:48.97530: fsm: term=1 index=25848 cmd=frames txn=25847 pages=2 commit=1 start
2018-10-14 21:37:48.97534: fsm: term=1 index=25848 cmd=frames txn=25847 pages=2 commit=1 unregister txn
2018-10-14 21:37:48.97535: fsm: term=1 index=25848 cmd=frames txn=25847 pages=2 commit=1 done
2018-10-14 21:37:48.97535: fsm: term=1 index=25849 cmd=frames txn=25848 pages=2 commit=1 start
2018-10-14 21:37:48.97538: fsm: term=1 index=25849 cmd=frames txn=25848 pages=2 commit=1 unregister txn
2018-10-14 21:37:48.97538: fsm: term=1 index=25849 cmd=frames txn=25848 pages=2 commit=1 done
2018-10-14 21:37:48.97541: fsm: term=1 index=25850 cmd=frames txn=25849 pages=3 commit=1 start
2018-10-14 21:37:48.97543: fsm: term=1 index=25850 cmd=frames txn=25849 pages=3 commit=1 unregister txn
2018-10-14 21:37:48.97543: fsm: term=1 index=25850 cmd=frames txn=25849 pages=3 commit=1 done
2018-10-14 21:37:48.97545: fsm: term=1 index=25851 cmd=frames txn=25850 pages=10 commit=1 start
2018-10-14 21:37:48.97551: fsm: term=1 index=25851 cmd=frames txn=25850 pages=10 commit=1 unregister txn
2018-10-14 21:37:48.97551: fsm: term=1 index=25851 cmd=frames txn=25850 pages=10 commit=1 done
2018-10-14 21:37:48.97551: fsm: term=1 index=25852 cmd=frames txn=25851 pages=2 commit=1 start
2018-10-14 21:37:48.97554: fsm: term=1 index=25852 cmd=frames txn=25851 pages=2 commit=1 unregister txn
2018-10-14 21:37:48.97554: fsm: term=1 index=25852 cmd=frames txn=25851 pages=2 commit=1 done
2018-10-14 21:37:48.97555: fsm: term=1 index=25853 cmd=frames txn=25852 pages=2 commit=1 start
2018-10-14 21:37:48.97557: fsm: term=1 index=25853 cmd=frames txn=25852 pages=2 commit=1 unregister txn
2018-10-14 21:37:48.97557: fsm: term=1 index=25853 cmd=frames txn=25852 pages=2 commit=1 done
2018-10-14 21:37:48.97559: fsm: term=1 index=25854 cmd=frames txn=25853 pages=3 commit=1 start
2018-10-14 21:37:48.97567: fsm: term=1 index=25854 cmd=frames txn=25853 pages=3 commit=1 unregister txn
2018-10-14 21:37:48.97567: fsm: term=1 index=25854 cmd=frames txn=25853 pages=3 commit=1 done
2018-10-14 21:37:48.97568: fsm: term=1 index=25855 cmd=frames txn=25854 pages=11 commit=1 start
2018-10-14 21:37:48.97575: fsm: term=1 index=25855 cmd=frames txn=25854 pages=11 commit=1 unregister txn
2018-10-14 21:37:48.97575: fsm: term=1 index=25855 cmd=frames txn=25854 pages=11 commit=1 done
2018-10-14 21:37:48.97575: fsm: term=1 index=25856 cmd=frames txn=25855 pages=3 commit=1 start
2018-10-14 21:37:48.97579: fsm: term=1 index=25856 cmd=frames txn=25855 pages=3 commit=1 unregister txn
2018-10-14 21:37:48.97580: fsm: term=1 index=25856 cmd=frames txn=25855 pages=3 commit=1 done
2018-10-14 21:37:48.97581: fsm: term=1 index=25857 cmd=frames txn=25856 pages=10 commit=1 start
2018-10-14 21:37:48.97588: fsm: term=1 index=25857 cmd=frames txn=25856 pages=10 commit=1 unregister txn
2018-10-14 21:37:48.97588: fsm: term=1 index=25857 cmd=frames txn=25856 pages=10 commit=1 done
2018-10-14 21:37:48.97589: fsm: term=1 index=25858 cmd=frames txn=25857 pages=9 commit=1 start
2018-10-14 21:37:48.97598: fsm: term=1 index=25858 cmd=frames txn=25857 pages=9 commit=1 unregister txn
2018-10-14 21:37:48.97598: fsm: term=1 index=25858 cmd=frames txn=25857 pages=9 commit=1 done
2018-10-14 21:37:48.97599: fsm: term=1 index=25859 cmd=checkpoint file=db.bin start
2018-10-14 21:37:48.97631: fsm: term=1 index=25859 cmd=checkpoint failed: checkpoint: disk I/O error


goroutine 55 [running]:
github.com/CanonicalLtd/go-dqlite/internal/trace.(*Tracer).Panic(0xc00031e000, 0x1159e31, 0x2, 0xc000cf7d68, 0x1, 0x1)
	/build/lxd/parts/lxd/go/src/github.com/CanonicalLtd/go-dqlite/internal/trace/tracer.go:59 +0x12a
github.com/CanonicalLtd/go-dqlite/internal/replication.(*FSM).Apply(0xc000180ee0, 0xc000e8aff0, 0x0, 0x0)
	/build/lxd/parts/lxd/go/src/github.com/CanonicalLtd/go-dqlite/internal/replication/fsm.go:84 +0x136
github.com/hashicorp/raft.(*Raft).runFSM.func1(0xc000e861c0)
	/build/lxd/parts/lxd/go/src/github.com/hashicorp/raft/fsm.go:57 +0x155
github.com/hashicorp/raft.(*Raft).runFSM(0xc00033c000)
	/build/lxd/parts/lxd/go/src/github.com/hashicorp/raft/fsm.go:120 +0x2ef
github.com/hashicorp/raft.(*Raft).runFSM-fm()
	/build/lxd/parts/lxd/go/src/github.com/hashicorp/raft/api.go:506 +0x2a
github.com/hashicorp/raft.(*raftState).goFunc.func1(0xc00033c000, 0xc000178630)
	/build/lxd/parts/lxd/go/src/github.com/hashicorp/raft/state.go:146 +0x53
created by github.com/hashicorp/raft.(*raftState).goFunc
	/build/lxd/parts/lxd/go/src/github.com/hashicorp/raft/state.go:144 +0x66
```

If there's any further information required or anything I can test, you can ping me on #lxcontainers for a faster response.

--- a/lxd/db/cluster/update.go
+++ b/lxd/db/cluster/update.go
@@ -1,8 +1,10 @@
 package cluster
 
 import (
+	"context"
 	"database/sql"
 	"fmt"
+	"time"
 
 	"github.com/lxc/lxd/lxd/db/query"
 	"github.com/lxc/lxd/lxd/db/schema"
@@ -77,6 +79,11 @@ func updateFromV11(tx *sql.Tx) error {
 	if err != nil {
 		return errors.Wrap(err, "Failed to count rows in current tables")
 	}
+
+	// Use a large timeout since the update might take a while, due to the
+	// new indexes being created.
+	ctx, cancel := context.WithTimeout(context.Background(), time.Minute)
+	defer cancel()
 
 	stmts := fmt.Sprintf(`
 CREATE TABLE projects (
@@ -390,7 +397,7 @@ CREATE INDEX images_project_id_idx ON images (project_id);
 CREATE INDEX images_aliases_project_id_idx ON images_aliases (project_id);
 CREATE INDEX profiles_project_id_idx ON profiles (project_id);
 `)
-	_, err = tx.Exec(stmts)
+	_, err = tx.ExecContext(ctx, stmts)
 	if err != nil {
 		return errors.Wrap(err, "Failed to add project_id column")
 	}

